src/python/pants/bsp/protocol.py

```diff
@@ -70,21 +92,34 @@ def __init__(
         self,
         scheduler_session: SchedulerSession,
         union_membership: UnionMembership,
+        context: BSPContext,
         inbound: BinaryIO,
         outbound: BinaryIO,
         max_workers: int = 5,
     ) -> None:
         self._scheduler_session = scheduler_session
         self._inbound = JsonRpcStreamReader(inbound)
         self._outbound = JsonRpcStreamWriter(outbound)
-        self._initialized = False
+        self._initialized_lock = RLock()
+        self._initialized_value = False
+        self._context: BSPContext = context
         self._endpoint = Endpoint(self, self._send_outbound_message, max_workers=max_workers)
 
         self._handler_mappings: dict[str, type[BSPHandlerMapping]] = {}
         impls = union_membership.get(BSPHandlerMapping)
         for impl in impls:
             self._handler_mappings[impl.method_name] = impl
 
+    @property
+    def _initialized(self) -> bool:
+        with self._initialized_lock:
+            return self._initialized_value
+
+    @_initialized.setter
+    def _initialized(self, new_value: bool) -> None:
+        with self._initialized_lock:
+            self._initialized_value = new_value
```

Comment:
It would be good to understand the concurrency situation for `_handle_inbound_message` to determine whether this is truly necessary. Does `pylsp_jsonrpc` execute multiple handlers concurrently?

If it's actually necessary to synchronize these, then it would be because it is being accesses from multiple threads working on messages for one connection. If that's the case, then there is likely more synchronization necessary: in particular, around the BSPContext.

Reply:
> It would be good to understand the concurrency situation for _handle_inbound_message to determine whether this is truly necessary. Does pylsp_jsonrpc execute multiple handlers concurrently?

Yes it does.

Reply:
Hm. Then might it be more self-contained to manage all locking internally to the `BSPContext`, perhaps? Its fields will need protection as well.

Reply:
Maybe if there were shared mutable state that had to be manipulated concurrently. Given that the only use of `BSPContext` in this PR is to store the client-provided initialization parameters, however, concurrent access will be to an immutable value that was setup outside of the engine under proper serialization.

In the BSP protocol, any methods calls or notifications sent before the `build/initialize` RPC is made by the client must be discarded. All other threads will send an error back to the client without entering the engine or accessing `BSPContext`. See the first few lines of `_handle_inbound_message`. Only the thread that receives `build/initialize` will update the `BSPContext` with the client-provided setup parameters, and then allow other threads to not error by updating `_initialized`to `True`.

If additional state is added to `BSPContext`, then that particular state might need protection. For example, I intend to add a method to send notifications to the client. That might need locking or not depending on what the protocol library we are using does. But that locking can be handled when the applicable state is added.

Reply:
Ok. I think that that all makes sense, and inlining some of the explanation would help. I still think that maybe isolating it to the `BSPContext` (in another file?) would be a good idea. We deal with concurrency concerns so infrequently in Python that isolating it and calling it out is worth the time.

Reply:
Maybe the initialization check should check `client_params` in `BSPContext` (with locking around updating it) instead of a separate initialization value? If it is `None`, then `_handle_inbound_message` should error and if it exists then allow processing. Otherwise, moving `BSPContext` to another file won't also move the initialization lock added to `BSPConnection` by this PR.

Reply:
> Maybe the initialization check should check `client_params` in `BSPContext` (with locking around updating it) instead of a separate initialization value?

Yea, reasonable. To be clear: I was suggesting a new file _in addition to_ https://github.com/pantsbuild/pants/pull/14583#discussion_r813439120

Reply:
👍

Reply:
Context is moved to `context.py` and all locking is handled there.